provider/pkg/provider/template_source.go

```diff
@@ -0,0 +1,280 @@
+	"fmt"
+	"path"
+	"strings"
+
+	pbempty "google.golang.org/protobuf/types/known/emptypb"
+
+	"github.com/pulumi/pulumi-pulumiservice/provider/pkg/internal/pulumiapi"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+type PulumiServiceTemplateSourceResource struct {
+	client *pulumiapi.Client
+}
+
+type PulumiServiceTemplateSourceDestination struct {
+	Url *string
+}
+
+type PulumiServiceTemplateSourceInput struct {
+	OrganizationName string
+	SourceName       string
+	SourceURL        string
+	Destination      *PulumiServiceTemplateSourceDestination
+}
+
+func (i *PulumiServiceTemplateSourceInput) ToPropertyMap() resource.PropertyMap {
+	pm := resource.PropertyMap{}
+	pm["organizationName"] = resource.NewPropertyValue(i.OrganizationName)
+	pm["sourceName"] = resource.NewPropertyValue(i.SourceName)
+	pm["sourceURL"] = resource.NewPropertyValue(i.SourceURL)
+	if i.Destination != nil {
+		destinationMap := resource.PropertyMap{}
+		if i.Destination.Url != nil {
+			destinationMap["url"] = resource.NewPropertyValue(i.Destination.Url)
+		}
+		pm["destination"] = resource.NewObjectProperty(destinationMap)
+	}
+	return pm
+}
+
+func (s *PulumiServiceTemplateSourceResource) ToPulumiServiceTemplateSourceInput(inputMap resource.PropertyMap) (*PulumiServiceTemplateSourceInput, error) {
+	input := PulumiServiceTemplateSourceInput{}
+
+	input.OrganizationName = inputMap["organizationName"].StringValue()
+	input.SourceName = inputMap["sourceName"].StringValue()
+	input.SourceURL = inputMap["sourceURL"].StringValue()
+
+	if inputMap["destination"].HasValue() && inputMap["destination"].IsObject() {
+		destinationMap := inputMap["destination"].ObjectValue()
+		destination := PulumiServiceTemplateSourceDestination{}
+		if destinationMap["url"].HasValue() && destinationMap["url"].IsString() {
+			value := destinationMap["url"].StringValue()
+			destination.Url = &value
+		}
+		input.Destination = &destination
+	}
+	return &input, nil
+}
+
+func (s *PulumiServiceTemplateSourceResource) Name() string {
+	return "pulumiservice:index:TemplateSource"
+}
+
+func (s *PulumiServiceTemplateSourceResource) Diff(req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
+	olds, err := plugin.UnmarshalProperties(req.GetOldInputs(), plugin.MarshalOptions{KeepUnknowns: false, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	diffs := olds.Diff(news)
+	if diffs == nil {
+		return &pulumirpc.DiffResponse{
+			Changes: pulumirpc.DiffResponse_DIFF_NONE,
+		}, nil
+	}
+
+	dd := plugin.NewDetailedDiffFromObjectDiff(diffs, false)
+
+	detailedDiffs := map[string]*pulumirpc.PropertyDiff{}
+	replaces := []string(nil)
+	replaceProperties := map[string]bool{
+		"organizationName": true,
+	}
```

Comment:
just for my own knowledge why do we always replace `organizationName`

Reply:
So if `organizationName` changes, we *can't* update the resource, like, a TemplateSource created in one org, cannot be moved into another org.
So we have to delete old one and make new one in new org

Reply:
ah ok - sorry the naming here threw me. These are properties that will trigger a replacement requirement no properties to replace.